presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,788 @@
+    @Override
+    public void destroy()
+    {
+        PendingRead pendingRead;
+        synchronized (this) {
+            if (state.setIf(FINISHED, oldState -> !oldState.isTerminal())) {
+                close();
+            }
+            pendingRead = this.pendingRead;
+            this.pendingRead = null;
+        }
+
+        if (pendingRead != null) {
+            pendingRead.completeResultFutureWithEmpty();
+        }
+    }
+
+    @Override
+    public void fail()
+    {
+        if (state.setIf(BufferState.FAILED, oldState -> !oldState.isTerminal())) {
+            finalizerService.addFinalizer(this, this::close);
+        }
+    }
+
+    private synchronized void close()
+    {
+        for (HandleInfo info : handleInfoQueue) {
+            info.removeFile();
+        }
+        pages.clear();
+        handleInfoQueue.clear();
+        noMorePages.set(true);
+        totalBufferedPages.set(0);
+        totalBufferedBytes.set(0);
+        totalPagesRemaining.set(0);
+    }
+
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        // NOOP
+    }
+
+    @Override
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        // NOOP
+    }
+
+    @Override
+    public boolean isFinishedForLifespan(Lifespan lifespan)
+    {
+        return isFinished();
+    }
+
+    private long getPagesSize(Collection<SerializedPage> pages)
+    {
+        return pages.stream().mapToLong(SerializedPage::getRetainedSizeInBytes).sum();
+    }
+
+    private long getPagesRows(Collection<SerializedPage> pages)
+    {
+        return pages.stream().mapToLong(SerializedPage::getPositionCount).sum();
+    }
+
+    private class HandleInfo
+    {
+        private final Range<Long> range;
+        private final ListenableFuture<TempStorageHandle> handleFuture;
+        private final long bytes;
+        private final int pageCount;
+
+        public HandleInfo(Range<Long> range, ListenableFuture<TempStorageHandle> handleFuture, long bytes, int pageCount)
+        {
+            this.range = requireNonNull(range, "range is null");
+            this.handleFuture = requireNonNull(handleFuture, "handleFuture is null");
+            this.bytes = bytes;
+            this.pageCount = pageCount;
+        }
+
+        public long getBytes()
+        {
+            return bytes;
+        }
+
+        public int getPageCount()
+        {
+            return pageCount;
+        }
+
+        public Range<Long> getRange()
+        {
+            return range;
+        }
+
+        public ListenableFuture<TempStorageHandle> getHandleFuture()
+        {
+            return handleFuture;
+        }
+
+        public void removeFile()
+        {
+            executor.execute(() -> {
+                try {
+                    tempStorage.remove(context, handleFuture.get());
+                }
+                catch (Exception e) {
+                    log.error(e, "Failed to remove file from TempStorage");
```

Comment:
Is there a chance for us to get the file that we could not delete? Would it be in exception?

Reply:
I believe it'll be rare for us to have a file we cannot delete. The plan is to use a fire-and-forget executor to delete the files, and rely on the garbage collecting mechanism of tempstorage if it fails.